crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+                let header_level = caps["level"].len();
+                self.pop_sections_to_level(header_level);
+
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let section = Section {
+                    // HEADER_RE can't match without a match for group 'title'.
+                    title: caps.name("title").unwrap().into(),
+                    level: header_level,
+                    parent_id: Some(*parent),
+                };
+
+                if self.current_section_files.is_some() {
+                    return Err(anyhow::anyhow!(
+                        "Header '{}' not valid inside a test case; parent '{}' has code files.",
+                        section.title,
+                        self.sections[*parent].title,
+                    ));
+                }
+
+                let section_id = self.sections.push(section);
+                self.stack.push(section_id);
+
+                self.current_section_files = None;
+            } else if let Some(caps) = self.scan(&CODE_RE) {
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let mut config: FxHashMap<&'s str, &'s str> = FxHashMap::default();
+
+                if let Some(config_match) = caps.name("config") {
+                    for item in config_match.as_str().split_whitespace() {
+                        let mut parts = item.split('=');
+                        let key = parts.next().unwrap();
+                        let Some(val) = parts.next() else {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        };
+                        if parts.next().is_some() {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        }
+                        config.insert(key, val);
+                    }
+                }
+
+                let path = config.get("path").unwrap_or(&"test.py");
+
+                self.files.push(File {
+                    path,
+                    section: *parent,
+                    // CODE_RE can't match without matches for 'lang' and 'code'.
+                    lang: caps.name("lang").unwrap().into(),
+                    code: caps.name("code").unwrap().into(),
+                });
+
+                if let Some(current_files) = &mut self.current_section_files {
+                    if current_files.contains(*path) {
+                        let current = &self.sections[*self.stack.last().unwrap()];
+                        return Err(anyhow::anyhow!(
+                            "Test `{}` has duplicate files named `{path}`.",
+                            current.title
+                        ));
+                    }
+                    current_files.insert(*path);
+                } else {
```

Comment:
You can directly use insert since an existing file aborts the parsing anyway
```suggestion
                    if let Some(existing) = current_files.insert(*path) {
                        let current = &self.sections[*self.stack.last().unwrap()];
                        return Err(anyhow::anyhow!(
                            "Test `{}` has duplicate files named `{path}`.",
                            current.title
                        ));
                    }
                } else {
```

Reply:
`HashSet::contains` returns a `bool`, not `Option<T>`, but this approach still works.